src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -0,0 +1,275 @@
+        private IImmutableDictionary<TopicPartition, TaskCompletionSource<Done>> _subSources = ImmutableDictionary<TopicPartition, TaskCompletionSource<Done>>.Empty;
+        
+        /// <summary>
+        /// Kafka has signalled these partitions are revoked, but some may be re-assigned just after revoking.
+        /// </summary>
+        private IImmutableSet<TopicPartition> _partitionsToRevoke = ImmutableHashSet<TopicPartition>.Empty;
+        
+
+        protected StageActor SourceActor { get; private set; }
+        protected IActorRef ConsumerActor { get; private set; }
+
+        /// <summary>
+        /// SubSourceLogic
+        /// </summary>
+        public SubSourceLogic(SourceShape<(TopicPartition, Source<TMessage, NotUsed>)> shape, ConsumerSettings<K, V> settings,
+                              IAutoSubscription subscription, Func<SubSourceLogic<K, V, TMessage>, IMessageBuilder<K, V, TMessage>> messageBuilderFactory,
+                              Option<Func<IImmutableSet<TopicPartition>, Task<IImmutableSet<TopicPartitionOffset>>>> getOffsetsOnAssign,
+                              Action<IImmutableSet<TopicPartition>> onRevoke, TaskCompletionSource<NotUsed> completion) 
+            : base(shape)
+        {
+            _shape = shape;
+            _settings = settings;
+            _subscription = subscription;
+            _messageBuilder = messageBuilderFactory(this);
+            _getOffsetsOnAssign = getOffsetsOnAssign;
+            _onRevoke = onRevoke;
+            _completion = completion;
+
+            _updatePendingPartitionsAndEmitSubSourcesCallback = GetAsyncCallback<IImmutableSet<TopicPartition>>(UpdatePendingPartitionsAndEmitSubSources);
+            _partitionAssignedCallback = GetAsyncCallback<IImmutableSet<TopicPartition>>(HandlePartitionsAssigned);
+            _partitionRevokedCallback = GetAsyncCallback<IImmutableSet<TopicPartitionOffset>>(HandlePartitionsRevoked);
+            _stageFailCallback = GetAsyncCallback<ConsumerFailed>(FailStage);
+            _subsourceCancelledCallback = GetAsyncCallback<(TopicPartition, Option<ConsumeResult<K, V>>)>(HandleSubsourceCancelled);
+            _subsourceStartedCallback = GetAsyncCallback<(TopicPartition, TaskCompletionSource<Done>)>(HandleSubsourceStarted);
+            
+            SetHandler(shape.Outlet, onPull: EmitSubSourcesForPendingPartitions, onDownstreamFinish: PerformShutdown);
+        }
+
+        public override void PostStop()
+        {
+            ConsumerActor.Tell(new KafkaConsumerActorMetadata.Internal.Stop(), SourceActor.Ref);
+            
+            OnShutdown();
+            
+            base.PostStop();
+        }
+
+        protected override void OnTimer(object timerKey)
+        {
+            if (timerKey is CloseRevokedPartitions)
+            {
+                Log.Debug("#{} Closing SubSources for revoked partitions: {}", _actorNumber, _partitionsToRevoke.JoinToString(", "));
```

Comment:
Need to add numbers to the format string here

Reply:
Got it